File: mobile/src/main/java/com/nomi/artwatch/ui/activity/MainActivity.java
package com.nomi.artwatch.ui.activity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.wearable.companion.WatchFaceCompanion;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.wearable.Asset;
import com.google.android.gms.wearable.PutDataMapRequest;
import com.google.android.gms.wearable.PutDataRequest;
import com.google.android.gms.wearable.Wearable;
import com.nomi.artwatch.R;
import com.nomi.artwatch.di.component.ActivityComponent;
import com.nomi.artwatch.model.LoginModel;
import com.nomi.artwatch.model.PostModel;
import com.nomi.artwatch.model.UserModel;
import com.nomi.artwatch.ui.view.ArtView;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import rx.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

/**
 * Created by Ryota Niinomi on 2015/11/04.
 */
public class MainActivity extends DrawerActivity {

    private static final String PATH_OF_GIF = "/gif";
    private static final String KEY_GIF = "gif";

    private GoogleApiClient mGoogleApiClient;
    private String mPeerId;

    @Inject
    UserModel mUserModel;
    @Inject
    PostModel mPostModel;
    @Inject
    LoginModel mLoginModel;

    @Bind(R.id.artView)
    ArtView mArtView;

    private GoogleApiClient.ConnectionCallbacks mGoogleConnectionCallback = new GoogleApiClient.ConnectionCallbacks() {
        @Override
        public void onConnected(Bundle connectionHint) {
            if (mPeerId != null) {
                Timber.d("Connected to wear.");
            }
        }

        @Override
        public void onConnectionSuspended(int cause) {
            // TODO：Handling
        }
    };

    private GoogleApiClient.OnConnectionFailedListener mGoogleConnectionFailedListener = new GoogleApiClient.OnConnectionFailedListener() {
        @Override
        public void onConnectionFailed(ConnectionResult connectionResult) {
            // TODO：Error handling
        }
    };

    protected int getLayout() {
        return R.layout.activity_main;
    }

    @Override
    protected void injectDependency(ActivityComponent component) {
        component.inject(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ButterKnife.bind(this);

        mPeerId = getIntent().getStringExtra(WatchFaceCompanion.EXTRA_PEER_ID);
        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addConnectionCallbacks(mGoogleConnectionCallback)
                .addOnConnectionFailedListener(mGoogleConnectionFailedListener)
                .addApi(Wearable.API)
                .build();

        if (mLoginModel.isAuthorized()) {
            // Already authorized.
            fetchGifPosts();

        } else {
            // Authorize now.
            authorize();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        mGoogleApiClient.connect();
    }

    @Override
    protected void onStop() {
        if (mGoogleApiClient != null && mGoogleApiClient.isConnected()) {
            mGoogleApiClient.disconnect();
        }
        super.onStop();
    }

    /**
     * Authorize to Tumblr.
     */
    private void authorize() {
        mLoginModel
                .login()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(authUrl -> {
                    Timber.d("Auth URL : " + authUrl);
                    startActivity(new Intent("android.intent.action.VIEW", Uri.parse(authUrl)));

                }, throwable -> {
                    Timber.w(throwable, throwable.getLocalizedMessage());
                });
    }

    /**
     * Fetch Gif posts from Tumblr.
     */
    private void fetchGifPosts() {
        mPostModel
                .getPhotoPost("ryotaniinomi")
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(items -> {
                    mArtView.init(items, this::onGifSelected);

                }, throwable -> {
                    Timber.w(throwable, throwable.getLocalizedMessage());
                });
    }

    /**
     * Send an request to change gif image with selected one.
     *
     * @param url
     */
    private void onGifSelected(String url) {
        if (mPeerId != null) {
            Glide.with(this)
                    .load(url)
                    .asGif()
                    .into(new SimpleTarget<GifDrawable>() {
                        @Override
                        public void onResourceReady(GifDrawable resource, GlideAnimation<? super GifDrawable> glideAnimation) {
                            // Convert GifDrawable to Asset.
                            Asset asset = createAssetFromDrawable(resource);

                            PutDataMapRequest dataMap = PutDataMapRequest.create(PATH_OF_GIF);
                            dataMap.getDataMap().putAsset(KEY_GIF, asset);
                            PutDataRequest request = dataMap.asPutDataRequest();

                            // Send the request.
                            Wearable.DataApi.putDataItem(mGoogleApiClient, request);

                            Toast.makeText(MainActivity.this, "changed", Toast.LENGTH_SHORT).show();
                        }
                    });
        }
    }

    /**
     * Create asset for sending the gif data to wear.
     *
     * @param drawable
     * @return
     */
    private Asset createAssetFromDrawable(GifDrawable drawable) {
        byte[] byteArray = drawable.getData();
        return Asset.createFromBytes(byteArray);
    }
}

File: mobile/src/main/java/com/nomi/artwatch/ui/activity/MainActivity.kt
package com.nomi.artwatch.ui.activity

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.TextView
import butterknife.bindView
import com.nomi.artwatch.GifUrlProvider
import com.nomi.artwatch.GifUrlProvider.Type
import com.nomi.artwatch.R
import com.nomi.artwatch.data.entity.Gif
import com.nomi.artwatch.di.component.ActivityComponent
import com.nomi.artwatch.model.PostModel
import com.nomi.artwatch.ui.adapter.binder.BlogAdapter
import com.nomi.artwatch.ui.util.SnackbarUtil
import com.nomi.artwatch.ui.view.ArtView
import com.tumblr.jumblr.types.Photo
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Action1
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by Ryota Niinomi on 2015/11/04.
 */
class MainActivity : DrawerActivity() {

    companion object {
        fun createIntent(context: Context) = Intent(context, MainActivity::class.java)
    }

    private var mBlogAdapter: BlogAdapter? = null

    @Inject
    lateinit var mPostModel: PostModel

    val mArtView: ArtView by bindView(R.id.art_view)
    val mEmptyView: TextView by bindView(R.id.empty_view)

    val mOnItemSelectedListener: AdapterView.OnItemSelectedListener = object : AdapterView.OnItemSelectedListener {
        /**
         * Change blog.
         */
        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
            val blog = mBlogAdapter?.getItem(position)
            mBlogModel.currentBlog = blog
            showGifs()
            setUserThumb(blog?.name ?: "")

            Timber.d("onItemSelected : " + blog?.name)
        }

        override fun onNothingSelected(parent: AdapterView<*>?) {
            Timber.d("onNothingSelected")
        }
    }

    override val layout: Int = R.layout.activity_main
    override val toolbarName: Int = R.string.home
    override val shouldShowSpinner: Boolean get() = true

    override fun injectDependency(component: ActivityComponent) {
        component.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mArtView.init(Action1 { photoSize -> onGifSelected(photoSize) })
        initBlogList()
    }

    /**
     * Authorize to Tumblr.
     */
    private fun authorize() {
        val subscription = mLoginModel
                .login()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    Timber.d("Auth URL : " + it)
                    startActivity(Intent("android.intent.action.VIEW", Uri.parse(it)))
                }, {
                    SnackbarUtil.showAlert(this, it.message ?: return@subscribe)
                })
        mSubscriptionsOnDestroy.add(subscription)
    }

    /**
     * Initialize blog list.
     */
    private fun initBlogList() {
        mBlogAdapter = BlogAdapter(this)
        val subscription = mUserModel
                .user
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (!it.blogs.isEmpty()) {
                        mBlogModel.currentBlog = it.blogs[0]
                    }
                    // AdapterView.OnItemSelectedListener#onItemSelected will be called and then,
                    // showGifs will be executed.
                    mBlogAdapter?.setDataSet(it.blogs)
                    mSpinner.adapter = mBlogAdapter

                }, {
                    SnackbarUtil.showAlert(this, it.message ?: return@subscribe)
                })
        mSubscriptionsOnDestroy.add(subscription)

        mSpinner.onItemSelectedListener = mOnItemSelectedListener
    }

    private fun showGifs() {
        if (mLoginModel.isAuthorized) {
            // Already authorized.
            fetchGifPosts()

        } else {
            // Authorize now.
            authorize()
        }
    }

    private fun toggleEmptyView(isEmpty: Boolean) {
        if (isEmpty) {
            mArtView.visibility = View.GONE
            mEmptyView.visibility = View.VISIBLE

        } else {
            mArtView.visibility = View.VISIBLE
            mEmptyView.visibility = View.GONE
        }
    }

    /**
     * Fetch Gif posts from Tumblr.
     */
    private fun fetchGifPosts() {
        mPostModel
                .getPhotoPost(mBlogModel.currentBlog?.name ?: return)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    mArtView.setDataSet(toGif(it))
                    toggleEmptyView(it.isEmpty())
                }, {
                    SnackbarUtil.showAlert(this, it.message ?: return@subscribe)
                })
                .apply { mSubscriptionsOnDestroy.add(this) }
    }

    /**
     * To Gif entity.
     */
    private fun toGif(list: List<Photo>): List<Gif> {
        return Observable
                .from(list)
                .map { Gif(GifUrlProvider.getUrl(it.sizes, Type.MOBILE), it.sizes, it.caption) }
                .toList()
                .toBlocking()
                .single()
    }
}
